regulation-worker/internal/delete/batch/batch.go

```diff
@@ -0,0 +1,379 @@
+	if err != nil {
+		return fmt.Errorf("error while opening statusTrackerFile file for uploading: %w", err)
+	}
+	defer file.Close()
+
+	_, err = b.FM.Upload(file)
+	if err != nil {
+		return fmt.Errorf("error while uploading statusTrackerFile file: %w", err)
+	}
+	b.mutex.Unlock()
+	os.Remove(cleanedFile)
+	return nil
+}
+
+//TODO: aws s3 ListObject allows listing of at max 1000 object at a time. So, implement paginatin.
+//Delete users corresponding to input userAttributes from a given batch destination
+func Delete(ctx context.Context, job model.Job, destConfig map[string]interface{}, destName string) error {
+	fmFactory := filemanager.FileManagerFactoryT{}
+	fm, err := fmFactory.New(&filemanager.SettingsT{
+		Provider: destName,
+		Config:   destConfig,
+	})
+	if err != nil {
+		return fmt.Errorf("error while creating file manager: %w", err)
+	}
+
+	dm, err := getDeleteManager(destName)
+	if err != nil {
+		return fmt.Errorf("failed to get appropriate deleteManager, %w", err)
+	}
+
+	batch := Batch{
+		FM: fm,
+		DM: dm,
+	}
+	defer batch.cleanup(destConfig["prefix"].(string))
+	files, exist, err := batch.listFilesAndCheckTrackerFile(ctx)
+	if err != nil {
+		return err
+	}
+	//if statusTracker.txt exists then read it & remove all those files name from above gzFilesObjects,
+	//since those files are already cleaned.
+	if exist {
+
+		err = batch.download(ctx, statusTrackerFile)
+		if err != nil {
+			return fmt.Errorf("error while downloading statusTrackerFile:%w", err)
+		}
+
+		data, err := os.ReadFile(statusTrackerFile)
+		if err != nil {
+			return fmt.Errorf("error while reading statusTracker.txt:%w", err)
+		}
+
+		cleanedFiles := strings.Split(string(data), "\n")
+		cleanedFiles = cleanedFiles[:len(cleanedFiles)-1]
+		files = removeCleanedFiles(files, cleanedFiles)
+
+	} else {
+		filePtr, err := os.Create(statusTrackerFile)
+		if err != nil {
+			return fmt.Errorf("error while creating statusTrackerFile:%w", err)
+		}
+		filePtr.Close()
+	}
+	g, gCtx := errgroup.WithContext(ctx)
+	goRoutineCount := make(chan bool, maxGoRoutine)
+	defer close(goRoutineCount)
+	for i := 0; i < len(files); i++ {
+		_i := i
+		goRoutineCount <- true
+		g.Go(func() error {
+			err = withExpBackoff(batch.download, gCtx, files[_i].Key)
+			if err != nil {
+				return err
+			}
+			fileNamePrefix := strings.Split(files[_i].Key, "/")
+			err := batch.delete(gCtx, job.UserAttributes, fileNamePrefix[len(fileNamePrefix)-1])
+			if err != nil {
+				return fmt.Errorf("error while deleting object, %w", err)
+			}
+			err = withExpBackoff(batch.upload, gCtx, files[_i].Key)
+			if err != nil {
+				return fmt.Errorf("error while uploading cleaned file, %w", err)
+			}
+			<-goRoutineCount
+			return nil
+		})
+	}
+	err = g.Wait()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *Batch) cleanup(prefix string) {
+
+	err := b.FM.DeleteObjects([]string{prefix + "/" + statusTrackerFile})
```

Comment:
better use [filepath.Join](https://pkg.go.dev/path/filepath#Join) to make it safe and portable.
```suggestion
	err := b.FM.DeleteObjects(filepath.Join(prefix, statusTrackerFile))
```